### PR TITLE
Allow megaparsec-8

### PR DESCRIPTION
**Pull request by sjakobi**
This also officially removes support for megaparsec-6.5 – the build had been broken before though.

### CODE REVIEW

**Comment by sjakobi** [comment]
Initially I tried to preserve compatibility with megaparsec-6.5.0, which @singpolyma enabled in https://github.com/dhall-lang/dhall-haskell/pull/939.

Then I noticed that `master` already doesn't build with 6.5.0:

```
src/Dhall/Parser.hs:31:25: error:
    Module ‘Text.Megaparsec’ does not export ‘ParseErrorBundle(..)’
   |
31 | import Text.Megaparsec (ParseErrorBundle(..), PosState(..))
   |                         ^^^^^^^^^^^^^^^^^^^^

src/Dhall/Parser.hs:31:47: error:
    Module ‘Text.Megaparsec’ does not export ‘PosState(..)’
   |
31 | import Text.Megaparsec (ParseErrorBundle(..), PosState(..))
   |   
```

Is compatibility with 6.5.0 an important goal? In that case I'd try to restore compatibility and test it in CI.

If we don't want to restore compatibility with 6.5.0, can support for megaparsec-7 be removed too? In that case we could remove even more CPP.

File: dhall/src/Dhall/Parser/Expression.hs [hunk 1]
```diff
@@ -61,12 +61,12 @@ getOffset = Text.Megaparsec.stateTokensProcessed <$> Text.Megaparsec.getParserSt
...
+setOffset o = Text.Megaparsec.updateParserState $ \state ->
+    state
 #if MIN_VERSION_megaparsec(7, 0, 0)
-setOffset o = Text.Megaparsec.updateParserState $ \(Text.Megaparsec.State s _ pst) ->
-  Text.Megaparsec.State s o pst
+        { Text.Megaparsec.stateOffset = o }
 #else
-setOffset o = Text.Megaparsec.updateParserState $ \(Text.Megaparsec.State s p _ stw) ->
-  Text.Megaparsec.State s p o stw
+        { Text.Megaparsec.stateTokensProcessed = o }
```
Comment (sjakobi) [inline dhall/src/Dhall/Parser/Expression.hs]: Did this increase laziness? Not sure whether we'd want that…

**Comment by sjakobi** [comment]
In case we drop compatibility with megaparsec-7, it seems that we could entirely remove the error position hack which we use `setOffset` for. See https://markkarpov.com/post/megaparsec-8.html#control-over-parse-error-locations.

**Comment by Gabriel439** [comment]
@sjakobi: Compatibility with `megaparsec-6.5.0` is no longer a goal.  Really, all that matters is that there was at least one `dhall` release that supported both newer and older versions of `megaparsec` (to ease the migration), but after that we only need to support newer versions of `megaparsec`

**Pull request merged**